Stacked chart interpolation

MarcelGerber opened the pull request:
Fixes the issues with (relative) stacked area charts outlined here: https://www.notion.so/owid/Relative-StackedArea-goes-beyond-100-465d2e80ccc14fc8b50a6e51768e3421

MarcelGerber commented on coreTable/CoreTable.ts:
```diff
@@ -70,6 +70,7 @@ const TransformsRequiringCompute = new Set([
     TransformType.AppendRows,
     TransformType.AppendColumns,
     TransformType.UpdateRows,
+    TransformType.UpdateColumnDefs,
```
@breck7 you made `toPercentageFromEachEntityForEachTime` column-based in https://github.com/owid/owid-grapher/commit/f591fe106f37efddf1834148ed1121d1a5661e89#diff-d0c27594f617591ec24495871c2d1eb77b60e4debc13c45614cae1fb16c887f7R303-R306, but didn't add `UpdateColumnDefs` to `TransformsRequiringCompute`. I hope this is the right thing to do?

danielgavrilov replied:
Yeah we were discussing this: https://owid.slack.com/archives/C01D295TM4K/p1603838545030400?thread_ts=1603830711.024300&cid=C01D295TM4K

But not sure what the right approach is!

breck7 replied:
Ugh, I really don't like `TransformsRequiringCompute` and think we should probably get rid of that ugly mess. I think what maybe will work better is this:

1. Get rid of `fn` on column definition. Instead transforms should just pass in a new transformed columnStore, or a `values` array on column defs, or a `transform` string which is basically a run-time DSL for describing transforms.
2. Then all 3 of those patterns would be column based and fast, and we can ditch TransformsRequiringCompute and just always build the columnStore in the constructor immediately (well, maybe we could lazily compute `transforms` if that's every a perf issue later on) .

breck7 replied:
So for now, if the tests pass and it doesn't throw perf off noticeably, this is fine. I think the only thing that may be worse perf wise here would be CovidExplorer v1, but hopefully we can nuke that code today or tomorrow.

MarcelGerber commented on grapher/stackedCharts/StackedAreaChart.tsx:
```diff
@@ -486,10 +482,6 @@ export class StackedAreaChart
     }
 
     @computed get series() {
-        if (this.props.disableLinearInterpolation === true)
```
Do we absolutely need `disableLinearInterpolation`?
I didn't re-implement it since interpolation is now applied in the constructor already...

breck7 replied:
I would say absolutely. I think it's very important for reproducibility/"show your work", and it also makes testing interpolation super easy:

![Screen Shot 2020-11-05 at 9 32 01 AM](https://user-images.githubusercontent.com/74692/98287666-c4d06f00-1f49-11eb-8154-755309ef6c33.png)
![Screen Shot 2020-11-05 at 9 31 58 AM](https://user-images.githubusercontent.com/74692/98287719-d9146c00-1f49-11eb-93d3-4287dc169392.png)

breck7 replied:
That being said, I don't think it matters where we turn it on or off, as long as it's easy to do at the story level.

MarcelGerber replied:
Nevermind, I added it back and it was easy. The story there might actually be useful.

MarcelGerber commented on grapher/stackedCharts/AbstractStackedChart.tsx:
```diff
@@ -36,6 +36,9 @@ export class AbstactStackedChart
             this.selectionArray.selectedEntityNames
         )
 
+        for (const slug of this.yColumnSlugs)
+            table = table.interpolateColumnsLinearly(slug)
```
This is applying the transform for each column, which may be an antipattern, idk?
But rewriting `interpolateColumnsLinearly` to work on multiple columns at once seems non-trivial, since the time column can differ for each column slug (I think?).

danielgavrilov replied:
Yeah I think this is fine, I do it for the tolerance interpolation as well for the same reason.

breck7 replied:
IIRC most other plotting libraries always use rows for series, and never columns, and would do a transpose to do that, so might simplify things for us if we do it that way too. (But I haven't double checked)

danielgavrilov replied:
I don't get it, I think the strategy doesn't matter here? Here's how I understand it, in both cases we are interpolating the same way, along time: 

<img width="1115" alt="Screenshot 2020-11-05 at 20 03 06" src="https://user-images.githubusercontent.com/1308115/98290719-f37a3400-1fa1-11eb-9dc8-7c00020599c1.png">

([spreadsheet](https://docs.google.com/spreadsheets/d/1uIwBAB76syIUYaUL6cr-5j2sSSzJVsMdp93Lg7bgBiM/edit?usp=sharing))

breck7 replied:
Awesome! You are totally right. I totally did the algebra wrong.

breck7 replied:
On the original topic, I think doing interpolation one at a time makes sense and is probably the better thing to do even if there was no time unit issue.

breck7 commented on coreTable/CoreTableUtils.ts:
```diff
@@ -158,11 +158,63 @@ export const makeRowFromColumnStore = (
+export function linearInterpolation(
+    valuesSortedByTimeAsc: (number | InvalidCell)[],
+    timesAsc: Time[],
+    start = 0,
```
@MarcelGerber I haven't looked myself, but is there any way we could share more code with the other interpolation? Or is it easier to have 2 totally separate methods?

@danielgavrilov what do you think of this pattern of passing in the starts and ends? It does ensure interpolation is O(N) (assuming things are passed in sorted), but also feels a little brittle in that it may be premature optimization?

Not entirely sure.

It feels like if we had fast groupBy, we wouldn't need it, but not sure if we could get that pattern fast enough?

Anyway, just thinking out loud.

danielgavrilov replied:
Seems fine to me, and probably as optimized as we can make this? 

The groupBy class/method would be cleaner to use but probably hard to achieve the same performance since we can't make the assumption in it that the array will stay the same, some rows might be inserted into a group by some transform, so would need to chunk the columnStore and create copies.

breck7 commented on coreTable/OwidTable.test.ts:
```diff
@@ -507,6 +507,95 @@ describe("time domain", () => {
 
+describe("linear interpolation", () => {
+    const table = new OwidTable(
+        `gdp,year,entityName,entityId,entityCode
```
nit: I think you can drop entityId and entityCode now from the test case. Those should no longer be required to create an OwidTable

breck7 commented on coreTable/OwidTable.test.ts:
```diff
@@ -507,6 +507,95 @@ describe("time domain", () => {
+    it("applies interpolation", () => {
+        const interpolatedTable = table.interpolateColumnsLinearly("gdp")
+
+        expect(interpolatedTable.rows).toEqual(
```
Nit: for this one I think a better assertion would be:

expect(interpolatedTable.get("gdp").allValues.toEqual([10, 0, ...])

breck7 replied:
That way you can ensure all values are correct and it would be easier for a reader to look at the CSV and then the array and understand the test.

MarcelGerber replied:
Ah that's nicer, thanks!

breck7 commented on coreTable/OwidTable.ts:
```diff
@@ -268,12 +269,11 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
-            map.set(time, map.get(time) + values[index])
-        })
+        timeValues.forEach((time, index) =>
+            map.set(time, (map.get(time) ?? 0) + values[index])
```
Nice! I like how you pulled off the oneliner here.

breck7 commented on coreTable/OwidTable.ts:
```diff
@@ -287,10 +287,12 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
-            ),
+            values: col.parsedValues.map((val, index) => {
+                const timeTotal = timeTotals.get(timeValues[index])
+                if (timeTotal === undefined || timeTotal === 0)
```
Hmm, timeTotal should never be undefined right? And the code proves that (at runtime anyway). I understand that when using a map TypeScript will give us that possibility, and that non-null assertions are bad, but is there a better way we could do this?

TypeScript gives an error for this, that `map.get(1)` is possibly undefined, even though we can easily see that it can never be undefined. So in this case better to use a non-null assertion than to handle the undefined cases?
```
const map = new Map<number, string>() 
map.set(1, "a")
const letterA = map.get(1).substr(0, 1)
```

MarcelGerber replied:
As you probably know by now, I like to play it safe and explicitly check for undefined, whereas you reach for type assertions more often :)

Yes, I think in this case we can use the type assertion. The risk here, as always, is when someone changes `sumsByTime` so it no longer returns the correct values in some case, and then this assumption that we have in our minds (but not in the type system!) no longer holds.
But there are many, many more (and more likely!) cases where this can happen, especially with array indices.

breck7 replied:
Ah, that's a very good point. This is a good thread about this type of situation: https://github.com/microsoft/TypeScript/issues/9619

breck7 replied:
But it does not seem like there is a solution yet

breck7 commented on coreTable/OwidTable.ts:
```diff
@@ -562,6 +564,62 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
         )
     }
 
+    interpolateColumnsLinearly(columnSlug: ColumnSlug) {
```
I don't know quite how, but it seems these interpolation functions could be broken up a bit into chains of smaller functions, and maybe share more code?

Any ideas @danielgavrilov ?

danielgavrilov replied:
Yeah it feels very similar, but there a few differences that might make the generalization tangled. E.g. the tolerance one distinguishes timeColumnOfTable and timeColumnOfValue, in order to be idempotent.

I'm not sure by looking at it – if Marcel wants to give it a go, great, but otherwise I think we can move onto other issues.

danielgavrilov replied:
Another thing is, `complete()` can inject a lot of unnecessary rows for tolerance implementation, it's the slowest part of the transform. We might want to optimize that and change the implementation.

MarcelGerber replied:
I tried my best, but as Daniel said it's really difficult.
One interpolation method alters one column, the other does two; one inserts a new column, the other doesn't; one differentiates between `timeColumnOfTable` and `timeColumnOfValue`, the other doesn't care -- all of these issues make it increasingly difficult to generalize here.
I still think a generalization and some thinking there would be very valuable because we really do have quite some duplicate code here. But let's move that until after the merge.

breck7 replied:
Sounds good!

MarcelGerber commented:
I addressed all the comments, and probably the biggest change I just pushed in f329b5f6d2a648fd116d270be93921909b0647d6 where I _attempted_ to generalize at least a little bit of the interpolation code.

From my side this is good to go now.

breck7 commented:
> I addressed all the comments, and probably the biggest change I just pushed in [f329b5f](https://github.com/owid/owid-grapher/commit/f329b5f6d2a648fd116d270be93921909b0647d6) where I _attempted_ to generalize at least a little bit of the interpolation code.
> 
> From my side this is good to go now.

Awesome! LGTM

danielgavrilov commented:
Nice! Clever generalization!

MarcelGerber commented:
Great, and thank you both for your awesome review! It has really been helpful this time.
